### PR TITLE
Check layer store for existing layers before fetching

### PR DESCRIPTION
**Pull request by jeffdh**
The implementation before this patch only checks for existing layers within the same image before downloading a layer.  However, on the storage side it de-duplicates the layers after already downloading the layer.  This adds a check for existing layers within the storage layer.  There is more complication that I don't understand yet as this feature was discussed in the [original PR](https://github.com/containers/image/pull/63#issuecomment-265253151), but for the scenario we download images this naively works.

This patch reads in the metadata on all layers once per image download and compares all potential layers to download before executing the download.

### CODE REVIEW

**Comment by jeffdh** [comment]
@aocole cc: @mffahey @apilloud

File: storage/storage_image.go [hunk 1]
```diff
@@ -285,6 +287,40 @@ func (s *storageImageDestination) PutBlob(stream io.Reader, blobinfo types.BlobI
...
 }
 
+// This is an expensive matching of a layer within the underlying storage
+func (s *storageImageDestination) findLayer(blobinfo types.BlobInfo) (*storage.Layer, int64, error) {
+
+	// Populate cache on first round
+	if s.cachedLayers == nil {
+		s.cachedLayers = make(map[string]*storage.Layer)
+		s.cachedSizes = make(map[string]int64)
+		layers, err := s.imageRef.transport.store.Layers()
```
Comment (aocole) [inline storage/storage_image.go]: Does `s.imageRef.transport.store.Layers()` get updated after each image is done downloading?

Comment (jeffdh) [reply storage/storage_image.go]: Yes.  See `storage.Store.PutLayer()` [here](https://github.com/containers/storage/blob/master/store.go#L786) calls `storage.LayerStore.Put()` which calls `storage.LayerStore.Save()` [here](https://github.com/containers/storage/blob/master/layers.go#L553).

**Comment by aocole** [comment]
LGTM

**Comment by jeffdh** [comment]
Reverted this PR

**Pull request merged**